File: spark/src/test/scala/geotrellis/spark/testfiles/SpaceTimeTestFileValues.scala
package geotrellis.spark.testfiles

import geotrellis.spark._
import geotrellis.raster._

import spire.syntax.cfor._

abstract class TestFileSpaceTimeTiles(tileLayout: TileLayout) {
  final def apply(key: SpaceTimeKey, timeIndex: Int): Tile = {
    val tile = FloatArrayTile.empty(tileLayout.tileCols, tileLayout.tileRows)

    cfor(0)(_ < tileLayout.tileRows, _ + 1) { row =>
      cfor(0)(_ < tileLayout.tileCols, _ + 1) { col =>
        tile.setDouble(col, row, value(key, timeIndex, col, row))
      }
    }

    tile
  }

  def value(key: SpaceTimeKey, timeIndex: Int, col: Int, row: Int): Double
}


class ConstantSpaceTimeTestTiles(tileLayout: TileLayout, v: Double) extends TestFileSpaceTimeTiles(tileLayout) {
  def value(key: SpaceTimeKey, timeIndex: Int, col: Int, row: Int): Double = v
}

/** Coordinates are CCC,RRR.TTT where C = column, R = row, T = time (year in 2010 + T).
  * So 34,025.004 would represent col 34, row 25, year 2014
  */
class CoordinateSpaceTimeTestTiles(tileLayout: TileLayout) extends TestFileSpaceTimeTiles(tileLayout) {
  def value(key: SpaceTimeKey, timeIndex: Int, col: Int, row: Int): Double= {
    val SpaceTimeKey(layoutCol, layoutRow, _) = key
    (layoutCol * 1000.0) + layoutRow + (timeIndex / 1000.0)
  }
}

File: geotrellis-python/tests/geotrellis_test/spark/testfiles/SpaceTimeTestFileValues.py
from __future__ import absolute_import
from geotrellis.raster.FloatArrayTile import FloatArrayTile

class TestFileSpaceTimeTiles(object):
    def __init__(self, tileLayout):
        self.tileLayout = tileLayout

    def __call__(self, key, timeIndex):
        tile = FloatArrayTile.empty(self.tileLayout.tileCols, self.tileLayout.tileRows)
        for row in xrange(0, self.tileLayout.tileRows):
            for col in xrange(0, self.tileLayout.tileCols):
                tile.setDouble(col, row, self.value(key, timeIndex, col, row))
        return tile

class ConstantSpaceTimeTestTiles(TestFileSpaceTimeTiles):
    def __init__(self, tileLayout, v):
        TestFileSpaceTimeTiles.__init__(self, tileLayout)
        self.v = v

    def value(self, key, timeIndex, col, row):
        return self.v

class CoordinateSpaceTimeTestTiles(TestFileSpaceTimeTiles):
    def __init__(self, tileLayout):
        TestFileSpaceTimeTiles.__init__(self, tileLayout)

    def value(self, key, timeIndex, col, row):
        layoutCol, layoutRow = key.col, key.row
        return (layoutCol * 1000.0) + layoutRow + (timeIndex / 1000.0)
